Migration

martinthomson opened the pull request:
This is the groundwork for migration.  I haven't done NEW_CONNECTION_ID yet (so migration doesn't work yet), but I wanted to put this up so that you can see where it is headed.

There are a few pieces of this that I want to extract and put in other pull requests as well.  The neqo-common stuff in particular can be separated.

The scary part of this change is that I ended up with a numeric reference to the current path.  This is hacktastic and I'm planning to store paths with `Rc<RefCell<Path>>` instead.  Then I can pass around a reference to the path instead; I can even use a weak reference in cases where the path isn't worth keeping (mostly related to closing).  That has some overheads involved with access, but it avoids the scary and brittle reference logic, which probably costs more anyway, both in terms of CPU cost and code cost.

One part of this design that I like is that datagrams are matched up with a path as they are received.  If it is a new path, then that path is "temporary".  If we like the packet, then we make the path a permanent one that we probe.  This eventually displaces old paths, as we only keep a few paths around.  Active paths don't go away and so it is the path that was active the longest time ago that gets the chop.  I like this design so I plan to keep that bit.

Next steps:
* build out NEW_CONNECTION_ID, which should be easy to do
* sending of probing packets needs special logic, because we don't want to commit too much to probing packets; right now, if a path wants to probe, then it gets all the frames
* the cleanup around path references
* cut this into manageable chunks
* and of course, going through and ensuring that all the important pieces have good test coverage

ddragana commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -1225,21 +1270,17 @@ impl Connection {
+
+                    qlog::packet_received(&mut self.qlog, &packet, &payload);
+                    let migrate = self.process_packet(path, &payload, now)?;
+                    self.postprocess(path, &packet, migrate, now)?;
```
Can we have a more descriptive name instead of postprocess?

ddragana reviewed (commented):
I did not get through all patches yet and I will need another pass through all patches.

These are just some comments.

martinthomson commented on neqo-transport/src/connection/tests/stream.rs:
```diff
@@ -299,30 +261,32 @@ fn simultaneous_stop_sending_and_reset() {
+    let ack = server.process(out.dgram(), now()).dgram();
 
-    let stream_readable = |e| matches!(e, ConnectionEvent::RecvStreamReadable {..});
+    let stream_readable =
+        |e| matches!(e, ConnectionEvent::RecvStreamReadable { stream_id: id } if id == stream_id);
```
With extra streams being created, extra filtering was needed here.

martinthomson commented on neqo-transport/src/connection/tests/stream.rs:
```diff
@@ -299,30 +261,32 @@ fn simultaneous_stop_sending_and_reset() {
+    client.stream_reset_send(stream_id, 0).unwrap();
+    let out_reset_frame = client.process(ack, now()).dgram();
+
+    // Send something out of order to force the server to generate an
+    // acknowledgment at the next opportunity.
```
The changed timing of acknowledgments was a little tricky here too.

martinthomson commented on neqo-transport/src/path.rs:
```diff
@@ -18,35 +33,447 @@ use neqo_common::Datagram;
+            .unwrap_or_else(|| Rc::new(RefCell::new(Path::temporary(local, remote))))
+    }
+
+    pub fn find_path_with_rebinding(&self, local: SocketAddr, remote: SocketAddr) -> PathRef {
```
Add a comment here.

martinthomson commented on neqo-transport/src/path.rs:
```diff
@@ -18,35 +33,447 @@ use neqo_common::Datagram;
+    /// This panics if `make_permanent` hasn't been called.
+    pub fn handle_migration(&mut self, path: &PathRef, remote: SocketAddr, now: Instant) {
+        qtrace!([self.primary().borrow()], "handle_migration");
+        for p in &self.paths {
+            qtrace!("Path {}", p.borrow());
+        }
```
```suggestion
```

martinthomson commented on neqo-transport/src/path.rs:
```diff
@@ -18,35 +33,447 @@ use neqo_common::Datagram;
+        if let Some(old_path) = self.select_primary(path) {
+            // Need to probe the old path if the peer migrates.
+            old_path.borrow_mut().probe();
+            // TODO(mt) - suppress probing if the path was valid within 3PTO.
```
I'm leaving this as a TODO because until the path owns the RTT estimator, it is difficult to plumb through the necessary state.

martinthomson commented on neqo-transport/src/path.rs:
```diff
@@ -106,4 +535,197 @@ impl Path {
+        } else if self.is_valid() {
+            usize::MAX
+        } else if let Some(limit) = self.received_bytes.checked_mul(3) {
+            let budget = if limit == 0 {
+                // If we have received absolutely nothing thus far, then this endpoint
+                // is the one initiating communication on this path.  Allow enough space for probing.
+                self.mtu() * 5
```
There's a gotcha here: we're going to probe any path we migrate to and send a padded datagram.  We should get 1200 bytes or more back.  If we don't, we're going to be bound by the amplification limit as a client.  This isn't ideal, but I think that it's tolerable.  We will ignore this amplification limit for any PTO that we send, so we aren't completely deadlocked.  It's a little annoying, that's all.

ddragana replied:
ok.

martinthomson commented on neqo-transport/src/recovery.rs:
```diff
@@ -581,13 +588,14 @@ impl PtoState {
+            Some(SendProfile::new_pto(self.space, mtu, self.probe))
         } else {
-            SendProfile::new_limited(0)
+            None
```
Note that this now returns `None` rather than returning a profile that only permits ACK frames.

This allows us to send probes on a path between the first and second PTO.  I found that we would not send anything at that time because - although we had plenty of congestion control window available - we were sending on PTO and therefore blocked.

This change might be a help if we have new data to send when the PTO timer is otherwise active.  This might be the source of some of our stalls.

martinthomson commented on neqo-transport/src/server.rs:
```diff
@@ -292,19 +300,22 @@ impl Server {
+                    }
                 } else {
-                    qerror!([self], "unable to encode retry, dropping packet");
                     None
```
Need to at least log an error here.  The generator shouldn't run dry.

martinthomson commented on neqo-transport/src/server.rs:
```diff
@@ -384,6 +395,30 @@ impl Server {
+            c.set_retry_cids(odcid, initial.src_cid, initial.dst_cid);
+        }
+        c.set_validation(Rc::clone(&self.address_validation));
+        if let Some(spa) = &self.preferred_address {
+            if c.set_preferred_address(spa).is_err() {
+                qwarn!([self], "Unable to set preferred address");
+            }
+        }
```
This is the new stuff.  The rest is just factored out.

martinthomson commented on neqo-transport/src/tparams.rs:
```diff
@@ -455,6 +587,232 @@ mod tests {
+        .encode(&mut Encoder::new(), PREFERRED_ADDRESS);
+    }
+
+    // pub fn new(v4: Option<SocketAddr>, v6: Option<SocketAddr>) -> Self {
+    //     assert!(v4.is_some() || v6.is_some());
+    //     if let Some(a) = v4 {
+    //         if let IpAddr::V4(addr) = a.ip() {
+    //             assert!(!addr.is_unspecified());
+    //         } else {
+    //             panic!("invalid address type for v4 address");
+    //         }
+    //         assert_ne!(a.port(), 0);
+    //     }
+    //     if let Some(a) = v6 {
+    //         if let IpAddr::V6(addr) = a.ip() {
+    //             assert!(!addr.is_unspecified());
+    //         } else {
+    //             panic!("invalid address type for v6 address");
+    //         }
+    //         assert_ne!(a.port(), 0);
+    //     }
+    //     Self { v4, v6 }
+    // }
+
```
```suggestion
```

ddragana reviewed (commented):
This is huge. 
It look fine and there is a lot of tests. 
Please address comment as you see fit and land it. 
Please do not land it before I create the next tag.

ddragana commented:
All the comments and commits messages were very useful, thanks.

martinthomson commented on neqo-transport/src/connection/tests/recovery.rs:
```diff
@@ -90,94 +86,65 @@ fn pto_works_full_cwnd() {
 fn pto_works_ping() {
     let mut client = default_client();
     let mut server = default_server();
     connect_force_idle(&mut client, &mut server);
```
Only those tests that count bytes (which are all in cc.rs) need to use that function.  It is harder to use `connect_rtt_idle` so it doesn't make sense to use it everywhere.

martinthomson commented on neqo-transport/src/path.rs:
```diff
@@ -36,12 +36,7 @@ const MAX_PATH_PROBES: usize = 3;
-    Temporary,
-    Id(u32),
-}
+pub type PathRef = Rc<RefCell<Path>>;
```
So do I.  It took me a little while to get there, but this seems to work pretty nicely.

martinthomson commented on neqo-transport/src/path.rs:
```diff
@@ -168,11 +180,37 @@ impl Paths {
+            builder.encode_varint(FRAME_TYPE_RETIRE_CONNECTION_ID);
+            builder.encode_varint(seqno);
+            tokens.push(RecoveryToken::RetireConnectionId(seqno));
+            stats.retire_connection_id += 1;
```
Good call.

martinthomson replied:
Hang on.  We generally count frames sent, including retransmissions.  Counting when we decide to retire would give us a different count as it could be some time until we put the frame out.  I'd rather get a precise accounting of the frames we send, and maybe look into counting losses so that we can get information on retransmissions - or remove them from counts - in a different way.

ddragana replied:
I think we need another stats that counts retired connectionId not frames.

martinthomson commented:
@ddragana can you give the [last few commits](https://github.com/mozilla/neqo/compare/0c5199f..cec7ebfcc6750ad8dc42932cee6efe40efd925f0) a sanity check?  I merged the latest code and moved the preferred address configuration to use ConnectionParameters.

In doing so, I integrated #1071 and then discovered that wasn't enough.  I needed to save the value of `ConnectionParameters` because we were not setting the stream limits properly in `StreamIndexes`.  So you probably shouldn't expect the current code on `main` to work with non-default values, especially if 0-RTT is rejected.

I've defaulted to having preferred addresses enabled at the client, but you can turn them off with `ConnectionParameters::disable_preferred_address`.

I've also moved the constants for initial stream limits to `ConnectionParameters`.  Code that needs to know the default value can inspect a default `ConnectionParameters`.

Oh, and I moved `StreamType` to the `stream_id` module.  I needed it for the new module, but it didn't make sense to reference `frame`.  I just made helper functions for `Frame` to use.

ddragana commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -350,15 +348,36 @@ impl Connection {
 
-        tps.set_integer(
+    /// Read connection parameters and update transport parameters.
+    fn read_parameters(&mut self) -> Res<()> {
```
I am wondering, shouldn't `read_parameters` return `Res<TransportParameters>` and be responsible for creating `TransportParameters`, setting its default and reading `self.conn_params`?
I do not like that there are 2 steps and in between `TransportParameters` are not correct.

martinthomson replied:
That's good feedback, but it's a fairly large restructuring.  Opened #1073 for that.  This is definitely large enough to warrant follow-up.

ddragana replied:
I agree, thanks.

ddragana reviewed (commented):
This looks good.
My comments re only suggestions, this is good for landing.

martinthomson commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -2543,7 +2554,10 @@ impl Connection {
-        self.indexes = StreamIndexes::new();
+        self.indexes = StreamIndexes::new(
+            self.conn_params.get_max_streams(StreamType::BiDi),
+            self.conn_params.get_max_streams(StreamType::UniDi),
```
Opened #1072.  If I get around to it before landing...
